### ISSUE TITLE
Randomly getting violation errors

### ISSUE THREAD

**Issue by ametrix** [UniversalRobots/Universal_Robots_ROS_Driver#485]
# Summary

When executing trajectories, we are getting random violation errors.

We are executing predefined trajectories, through scaled_pos_trajectory
_controller ROS action interface. Trying to move robot with speeds close to maximum speeds from specification: 
- Velocity for base and shoulder joints: Max 120 °/s. All other joints: Max 180 °/s. 
- Accelerations 360 °/s for all joints.
- Payload is set to: COG:[x:0.010, y:-0.025, z:0.060], MASS:2.060 kg

Our trajectories are generated with OMPL planner.

Generally we successfully move the robot, but randomly this error appears in teach pendant:
**C283A26 : Safety system: Force limitation: A joint exceeded the torque window by -0.07159 Nm**
**Explanation:** 
**If the target trajectory is very jerky, eg in case of sensor-based control with path_offset() or servoj(), the target joint torques can exceed safety system limits.** 
**Suggestion:** 
**Check that the target trajectory is smooth.  Noisy sensors or communication jitters in external control can cause jerky trajectories.**

Things we already checked:
- Changed the safety limits (from teach pendant to least restrictive. This eliminated violations, but this isn’t feasible for the current project)
- Checked safety planes (didn't help with violations)

- Updated the firmware (5.9, 5.10, 5.11 but didn't help) 

- Played with payload values (didn't help with violations)

- Reduced acceleration multiplier (lowering this to 0.25, dramatically reduces violations compared to 0.7 - 0.8, though some violations have still been observed at 0.6)

- Changed interpolation (Switching from quintic to cubic interpolation had significant impact on the frequency when testing with fw 5.9, though after upgrading to 5.11 the result was not conclusive. Switching to cubic also has the affect of reducing “loopy” motions around transitions which is desirable overall)

We don’t use a real time operating system, and so our controller updates are not fixed.  However, we have switched to using high priority threads which has dramatically improved the consistency of our updates.  Unfortunately this did not lead to an observable reduction in violations.

# Versions
 - ROS Driver version: (commit hash 44858fb - Fix package name in calibration example)
 - Affected Robot Software Version(s):  5.11, 5.10, 5.9
 - Affected Robot Hardware Version(s): ur10e
 - Robot Serial Number:
 - UR+ product(s) installed:
 - URCaps Software version(s): 1.0.5

# Impact
When executing trajectories, our robot stops randomly with violation errors.

## Use Case and Setup
- Using Universal Robots ROS Driver through controller interface.
- Running the driver without realtime kernel.
- Executing predefined trajectories generated with planner.

## Project status at point of discovered
When increased robot speed to more than 50% of maximum values from specs. 

## Expected Behavior
Execution trajectories without violation error.

## Actual Behavior
Executed trajectories aborted with violation error.

**Comment by ametrix** [comment]
Hey @urrsk! Thanks for your reply. I'm using most restricted settings and when I switch them to least restricted the violations are gone. Fortunately I found solution to my problem and there are 0 violations now. To fix these violation errors I had to expose servoj gain and servoj lookahead time. Looks like they are set by default to values which makes trajectory very sharp and robot tries to follow trajectory very aggressively. I found this by analyzing report generated when violation occurred. I have plotted the joint accelerations at the moment of violation. There were very big acceleration spikes on all joints. Setting exposed parameters to these solved the problem for me:

 UR_SERVOJ_GAIN    = "1000"
 UR_SERVOJ_LOOKAHEAD_TIME = "0.04" # (sec)
